arwen/elrondapi/bigFloatOps.go

```diff
@@ -0,0 +1,893 @@
+		return
+	}
+	managedType.ConsumeGasForBigFloatCopy(op1, op2, dest)
+
+	op1BigInt := big.NewInt(0)
+	op1.Int(op1BigInt)
+	op2BigInt := big.NewInt(0)
+	op1.Int(op2BigInt)
+	if op1BigInt.Sign() > 0 {
+		op1BigInt.Add(op1BigInt, big.NewInt(1))
+	}
+	if op2BigInt.Sign() > 0 {
+		op2BigInt.Add(op2BigInt, big.NewInt(1))
+	}
+
+	//this calculates the length of the result in bytes
+	lengthOfResult := big.NewInt(0).Div(big.NewInt(0).Mul(op2BigInt, big.NewInt(int64(op1BigInt.BitLen()))), big.NewInt(8))
+	managedType.ConsumeGasForThisBigIntNumberOfBytes(lengthOfResult)
+
+	dest.Set(pow(op1, op2))
+}
+
+func pow(base *big.Float, exp *big.Float) *big.Float {
+	result := new(big.Float).Copy(base)
+	counter := big.NewFloat(1)
+	basePrecision := base.Prec()
+	expPrecision := exp.Prec()
+	if basePrecision > expPrecision {
+		counter.SetPrec(basePrecision)
+		result.SetPrec(basePrecision)
+	} else {
+		counter.SetPrec(expPrecision)
+		result.SetPrec(expPrecision)
+	}
+	for counter.Cmp(exp) < 0 {
+		result.Mul(result, base)
+		counter.Add(counter, big.NewFloat(1))
+	}
+	return result
+}
+
+//export v1_4_bigFloatFloor
+func v1_4_bigFloatFloor(context unsafe.Pointer, opHandle, bigIntHandle int32) {
+	managedType := arwen.GetManagedTypesContext(context)
+	metering := arwen.GetMeteringContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+
+	gasToUse := metering.GasSchedule().BigFloatAPICost.BigFloatFloor
+	metering.UseGas(gasToUse)
+
+	op, err := managedType.GetBigFloat(opHandle)
+	bigIntOp := managedType.GetBigIntOrCreate(bigIntHandle)
+	if arwen.WithFault(err, context, runtime.BigFloatAPIErrorShouldFailExecution()) {
+		return
+	}
+	if oneIsInfinity(op) {
+		_ = arwen.WithFault(arwen.ErrInfinityFloatOperation, context, runtime.BigFloatAPIErrorShouldFailExecution())
+		return
+	}
+	managedType.ConsumeGasForBigFloatCopy(op)
+	op.Int(bigIntOp)
+	managedType.ConsumeGasForBigIntCopy(bigIntOp)
+	if op.IsInt() {
+		return
+	}
+	if bigIntOp.Sign() < 0 {
+		bigIntOp.Sub(bigIntOp, big.NewInt(1))
+	}
+}
+
+//export v1_4_bigFloatCeil
+func v1_4_bigFloatCeil(context unsafe.Pointer, opHandle, bigIntHandle int32) {
+	managedType := arwen.GetManagedTypesContext(context)
+	metering := arwen.GetMeteringContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+
+	gasToUse := metering.GasSchedule().BigFloatAPICost.BigFloatCeil
+	metering.UseGas(gasToUse)
+
+	op, err := managedType.GetBigFloat(opHandle)
+	bigIntOp := managedType.GetBigIntOrCreate(bigIntHandle)
+	if arwen.WithFault(err, context, runtime.BigFloatAPIErrorShouldFailExecution()) {
+		return
+	}
+	if oneIsInfinity(op) {
+		_ = arwen.WithFault(arwen.ErrInfinityFloatOperation, context, runtime.BigFloatAPIErrorShouldFailExecution())
+		return
+	}
+	managedType.ConsumeGasForBigFloatCopy(op)
+	op.Int(bigIntOp)
+	managedType.ConsumeGasForBigIntCopy(bigIntOp)
+	if op.IsInt() {
+		return
+	}
+	if bigIntOp.Sign() > 0 {
```

Comment:
Isn't here an edge case where the int is actually equal to the float?
ceil(1) = 1
ceil(1.1) = 2

Reply:
same as floor